src/common/device_helpers.cuh

```diff
@@ -402,57 +403,69 @@ struct XGBDefaultDeviceAllocatorImpl : XGBBaseDeviceAllocator<T> {
- * allocations if verbose. Does not initialise memory on construction.
+ * \brief Caching memory allocator, uses cub::CachingDeviceAllocator as a back-end, unless
+ *        RMM pool allocator is enabled.
  */
 template <class T>
-struct XGBCachingDeviceAllocatorImpl : thrust::device_malloc_allocator<T> {
+struct XGBCachingDeviceAllocatorImpl : XGBBaseDeviceAllocator<T> {
+  using SuperT = XGBBaseDeviceAllocator<T>;
   using pointer = thrust::device_ptr<T>;  // NOLINT
   template<typename U>
   struct rebind  // NOLINT
   {
     using other = XGBCachingDeviceAllocatorImpl<U>;  // NOLINT
   };
-   cub::CachingDeviceAllocator& GetGlobalCachingAllocator ()
-   {
-     // Configure allocator with maximum cached bin size of ~1GB and no limit on
-     // maximum cached bytes
-     static cub::CachingDeviceAllocator *allocator = new cub::CachingDeviceAllocator(2, 9, 29);
-     return *allocator;
-   }
-   pointer allocate(size_t n) {  // NOLINT
-     T *ptr;
-     GetGlobalCachingAllocator().DeviceAllocate(reinterpret_cast<void **>(&ptr),
-                                                n * sizeof(T));
-     pointer thrust_ptr(ptr);
-     GlobalMemoryLogger().RegisterAllocation(thrust_ptr.get(), n * sizeof(T));
-     return thrust_ptr;
-   }
-   void deallocate(pointer ptr, size_t n) {  // NOLINT
-     GlobalMemoryLogger().RegisterDeallocation(ptr.get(), n * sizeof(T));
-     GetGlobalCachingAllocator().DeviceFree(ptr.get());
-   }
-
-  __host__ __device__
-    void construct(T *)  // NOLINT
-  {
-    // no-op
+  cub::CachingDeviceAllocator& GetGlobalCachingAllocator() {
+    // Configure allocator with maximum cached bin size of ~1GB and no limit on
+    // maximum cached bytes
+    static cub::CachingDeviceAllocator *allocator = new cub::CachingDeviceAllocator(2, 9, 29);
+    return *allocator;
+  }
+  pointer allocate(size_t n) {  // NOLINT
+    pointer ptr{nullptr};
+    if (use_cub_allocator_) {
+      T* raw_ptr{nullptr};
+      GetGlobalCachingAllocator().DeviceAllocate(reinterpret_cast<void**>(&raw_ptr), n * sizeof(T));
+      ptr = pointer(raw_ptr);
+    } else {
+      ptr = SuperT::allocate(n);
+    }
+    GlobalMemoryLogger().RegisterAllocation(ptr.get(), n * sizeof(T));
+    return ptr;
+  }
+  void deallocate(pointer ptr, size_t n) {  // NOLINT
+    GlobalMemoryLogger().RegisterDeallocation(ptr.get(), n * sizeof(T));
+    if (use_cub_allocator_) {
+      GetGlobalCachingAllocator().DeviceFree(ptr.get());
+    } else {
+      SuperT::deallocate(ptr, n);
+    }
+  }
+#if defined(XGBOOST_USE_RMM) && XGBOOST_USE_RMM == 1
+  XGBCachingDeviceAllocatorImpl()
+      : SuperT(rmm::mr::get_current_device_resource(), cudaStream_t{nullptr}),
+        use_cub_allocator_(true) {
+    char* symbol = abi::__cxa_demangle(typeid(*SuperT::resource()).name(),
+                                       nullptr, nullptr, nullptr);
+    CHECK(symbol);
+    std::string t{symbol};
+    if (t.find("pool_memory_resource") != std::string::npos) {
```

Comment:
It may cause your users frustration if "pool_memory_resource" is the only resource you recognize if they have carefully constructed a resource that works better than a pool_memory_resource for their workflow.  RMM has composable resources, so for example a user might be using a `logging_resource_adapter` wrapped around a `binning_memory_resource` (or upcoming `arena_memory_resource`) with multiple different other resources in use for its bins.

Reply:
@harrism Do `binning_memory_resource` or `arena_memory_resource` allocate a large chunk of memory ahead of time, like `pool_memory_resource`?

Reply:
```suggestion
    if (t.find("pool_memory_resource") != std::string::npos || t.find("binning_memory_resource") != std::string::npos
          || t.find("arena_memory_resource") != std::string::npos ) {
```

@harrism We can exhaustively list all the allocators that have comparable performance as a pool allocator. It suffices that the full type string includes any of them as substring.

Reply:
They are fast suballocators (depending on the pattern, faster than pool_mr), if that's what you are asking. 

My point is that you cannot exhaustively list them all, because they are composable.

Reply:
> you cannot exhaustively list them all, because they are composable.

The symbol string is a fully qualified class name, such as:
```
rmm::mr::logging_resource_adapter<rmm::mr:: binning_memory_resource>
```
so checking for the substring `binning_memory_resource` will trigger the condition.